### PR TITLE
fix (fixing an issue): migrate to typescript

### PR DESCRIPTION
**Pull request by dependentmadani**
Hello,
I need a review of code for fixing issue #106. Thanks

### CODE REVIEW

File: package-lock.json
Comment (neSpecc) [inline package-lock.json:1]: this file is redundant since yarn is used in this repo

File: package.json [hunk 1]
```diff
@@ -30,6 +31,11 @@
     "email": "team@codex.so"
   },
   "devDependencies": {
+    "@rollup/plugin-typescript": "^11.1.6",
```
Comment (neSpecc) [inline package.json:34]: lets use "vite-plugin-dts" instead

File: package.json [hunk 2]
```diff
@@ -13,8 +13,9 @@
   "files": [
     "dist"
   ],
-  "main": "./dist/header.umd.js",
+  "main": "./dist/index.js",
```
Comment (neSpecc) [inline package.json:16]: unexpected change

File: package.json [hunk 3]
```diff
@@ -30,6 +31,11 @@
     "email": "team@codex.so"
   },
   "devDependencies": {
+    "@rollup/plugin-typescript": "^11.1.6",
+    "@types/node": "^20.14.2",
+    "ts-node": "^10.9.2",
+    "tslib": "^2.6.3",
```
Comment (neSpecc) [inline package.json:37]: Seems like these packages are not needed

File: src/index.ts [hunk 4]
```diff
@@ -11,6 +11,10 @@ import { IconH1, IconH2, IconH3, IconH4, IconH5, IconH6, IconHeading } from '@co
  * @property {string} text — Header's content
  * @property {number} level - Header's level from 1 to 6
  */
+interface HeaderData {
+  text: string;
+  level: number;
```
Comment (neSpecc) [inline src/index.ts:16]: please add jsdocs to each property. You can move them from the `@typedef` section which is not needed anymore

File: src/index.ts [hunk 5]
```diff
@@ -38,7 +120,15 @@ export default class Header {
...
    *   readOnly - read only mode flag
    */
-  constructor({ data, config, api, readOnly }) {
+  
+  private api: any;
+  private readOnly: boolean;
+  private _CSS: ToolSettings;
+  private _settings: HeaderConfig;
+  private _data: HeaderData;
+  private _element: HTMLHeadingElement;
```
Comment (neSpecc) [inline src/index.ts:129]: add docs please to each property

File: src/index.ts [hunk 6]
```diff
@@ -11,6 +11,10 @@ import { IconH1, IconH2, IconH3, IconH4, IconH5, IconH6, IconHeading } from '@co
  * @property {string} text — Header's content
  * @property {number} level - Header's level from 1 to 6
  */
+interface HeaderData {
```
Comment (neSpecc) [inline src/index.ts:14]: lets export this interface, it can be useful

File: src/index.ts [hunk 7]
```diff
@@ -19,6 +23,84 @@ import { IconH1, IconH2, IconH3, IconH4, IconH5, IconH6, IconHeading } from '@co
...
+}
+
+/**
+ * 
+ * @typedef {object} PasteEvent
+ * @description Event triggered on paste
+ * @property {object} detail - Event details
+ * @property {HTMLElement} detail.data - Pasted content
+ */
+interface PasteEvent {
```
Comment (neSpecc) [inline src/index.ts:53]: I think, it can be imported from Editor.js types

File: src/index.ts [hunk 8]
```diff
@@ -19,6 +23,84 @@ import { IconH1, IconH2, IconH3, IconH4, IconH5, IconH6, IconHeading } from '@co
...
+ * 
+ * @typedef {object} ToolSettings
+ * @description Tool's styles configuration
+ * @property {string} block - Block style
+ * @property {string} wrapper - Wrapper style
+ */
+interface ToolSettings {
+  block: string;
+  wrapper: string;
+}
```
Comment (neSpecc) [inline src/index.ts:69]: seems redundant, just move object type to a CSS getter

File: src/index.ts [hunk 9]
```diff
@@ -19,6 +23,84 @@ import { IconH1, IconH2, IconH3, IconH4, IconH5, IconH6, IconHeading } from '@co
...
+ * 
+ * @typedef {object} BlockTune
+ * @description Configuration for block tunes
+ * @property {string} icon - Icon for the tune
+ * @property {string} label - Label for the tune
+ * @property {function} onActivate - Function to call on activation
+ * @property {boolean} closeOnActivate - Whether to close on activation
+ * @property {boolean} isActive - Whether the tune is active
+ */
+interface BlockTune {
```
Comment (neSpecc) [inline src/index.ts:81]: can be also imported from editor js

File: src/index.ts [hunk 10]
```diff
@@ -395,7 +485,7 @@ export default class Header {
     ];
 
     return this._settings.levels ? availableLevels.filter(
-      l => this._settings.levels.includes(l.number)
+      l => this._settings.levels?.includes(l.number)
```
Comment (neSpecc) [inline src/index.ts:488]: ? seems unnecessary, since you're checking for emptiness before it

File: tsconfig.json
Comment (neSpecc) [inline tsconfig.json:1]: lets remove unused props

File: src/index.ts (hunk 10 above)
Comment (dependentmadani) [reply src/index.ts]: replace it with '!' to insure the non-null assertion, as it will not be build

**Comment by dependentmadani** [comment]
## Changes made:
- Remove the package-lock.json.
- Use `vite-plugin-dts` instead of rollup plugin.
- Add Jsdocs to each property.
- Export the HeaderData interface.
- Move object type into a CSS getter.
- Import necessary interfaces from editorjs.
- Remove the unused props from tsconfig file.

File: src/index.ts [hunk 11]
```diff
@@ -4,21 +4,55 @@
...
+}
 
 /**
- * @typedef {object} HeaderConfig
  * @description Tool's config from Editor
- * @property {string} placeholder — Block's placeholder
- * @property {number[]} levels — Heading levels
- * @property {number} defaultLevel — default level
  */
+interface HeaderConfig {
```
Comment (neSpecc) [inline src/index.ts:22]: Could be also used outside, do lets add export
```suggestion
export interface HeaderConfig {
```

**Review by neSpecc — DISMISSED** [review]

**Comment by dependentmadani** [comment]
## Change made:
- export the HeaderConfig interface.

**Review by neSpecc — APPROVED** [review]

**Pull request merged**